tiledb/api/src/group.rs

```diff
@@ -0,0 +1,386 @@
+pub type QueryType = crate::array::Mode;
+
+pub struct GroupInfo {
+    pub uri: String,
+    pub group_type: ObjectType,
+    pub name: String,
+}
+
+pub(crate) struct RawGroup {
+    ffi: *mut ffi::tiledb_group_t,
+}
+
+impl RawGroup {
+    pub fn new(ffi: *mut ffi::tiledb_group_t) -> Self {
+        RawGroup { ffi }
+    }
+}
+
+impl Deref for RawGroup {
+    type Target = *mut ffi::tiledb_group_t;
+    fn deref(&self) -> &Self::Target {
+        &self.ffi
+    }
+}
+
+impl Drop for RawGroup {
+    fn drop(&mut self) {
+        unsafe { ffi::tiledb_group_free(&mut self.ffi) }
+    }
+}
+
+pub struct Group<'ctx> {
+    context: &'ctx Context,
+    raw: RawGroup,
+}
+
+impl<'ctx> ContextBound<'ctx> for Group<'ctx> {
+    fn context(&self) -> &'ctx Context {
+        self.context
+    }
+}
+
+impl<'ctx> Group<'ctx> {
+    pub(crate) fn capi(&self) -> *mut ffi::tiledb_group_t {
+        *self.raw
+    }
+
+    pub(crate) fn new(context: &'ctx Context, raw: RawGroup) -> Self {
+        Group { context, raw }
+    }
+
+    pub fn create<S>(context: &'ctx Context, name: S) -> TileDBResult<()>
+    where
+        S: AsRef<str>,
+    {
+        let c_name = cstring!(name.as_ref());
+        context.capi_return(unsafe {
+            ffi::tiledb_group_create(context.capi(), c_name.as_ptr())
+        })
+    }
+
+    pub fn open<S>(
+        context: &'ctx Context,
+        uri: S,
+        query_type: QueryType,
+    ) -> TileDBResult<Self>
+    where
+        S: AsRef<str>,
+    {
+        let ctx = context.capi();
+        let mut group_raw: *mut ffi::tiledb_group_t = std::ptr::null_mut();
+
+        let c_uri = cstring!(uri.as_ref());
+
+        context.capi_return(unsafe {
+            ffi::tiledb_group_alloc(ctx, c_uri.as_ptr(), &mut group_raw)
+        })?;
+
+        let query_type_raw = query_type.capi_enum();
+        context.capi_return(unsafe {
+            ffi::tiledb_group_open(ctx, group_raw, query_type_raw)
+        })?;
+        Ok(Self::new(context, RawGroup::new(group_raw)))
+    }
+
+    pub fn uri(&self, context: &'ctx Context) -> TileDBResult<String> {
+        let c_context = self.context.capi();
+        let mut c_uri = std::ptr::null::<std::ffi::c_char>();
+        context.capi_return(unsafe {
+            ffi::tiledb_group_get_uri(c_context, Self::capi(self), &mut c_uri)
+        })?;
+        let uri = unsafe { std::ffi::CStr::from_ptr(c_uri) };
+        Ok(String::from(uri.to_string_lossy()))
+    }
+
+    pub fn query_type(
+        &self,
+        context: &'ctx Context,
+    ) -> TileDBResult<QueryType> {
+        let c_context = self.context.capi();
+        let mut c_type: ffi::tiledb_query_type_t = out_ptr!();
+        context.capi_return(unsafe {
+            ffi::tiledb_group_get_query_type(
+                c_context,
+                Self::capi(self),
+                &mut c_type,
+            )
+        })?;
+        QueryType::try_from(c_type)
+    }
+
+    pub fn delete_group<S>(
```

Comment:
Reading a bit more stuff, I now believe that this method fully deletes the entire group - is that correct?

If so then none of the methods would be usable afterwards, so this ought to have a consuming `self` rather than `&self`